### PR TITLE
ws: support cors

### LINKED ISSUE #436: WebSockets: add CORS support
CORS is not configured for WS handler, which causes problems with Ignite CLI.

### PR DESCRIPTION
**Pull request by tuxcanfly**
The websocket JSON RPC server did not support CORS. This PR implements this. Fixes #436 #742

### CODE REVIEW

File: rpc/json/ws.go
Comment (gupadhyaya) [inline rpc/json/ws.go:66]: what does this TODO mean? in future this will be changed?

Comment (tuxcanfly) [reply rpc/json/ws.go:66]: The config has these fields, we can either strictly follow the field names and only allow certain methods or headers.

But I looked at tendermint and they are basically just skipping these config fields, they're not used at all, see: https://github.com/cometbft/cometbft/blob/b0d64fa2f27d2439e655338117eb4eaa250a156c/rpc/jsonrpc/server/ws_handler.go#L49-L59

I'd say that tendermint config is actually misleading as people might think their RPC is gated by certain methods/headers while it's not. It's not a big security issue but still something that can catch off guard.

File: rpc/json/ws.go
Comment (gupadhyaya) [inline rpc/json/ws.go:68]: also, not need for many function calls and checks. you could inline something like (if it is sufficeint):
```
if h.config.IsCorsEnabled() {
	upgrader.CheckOrigin = func(r *http.Request) bool {
		for _, allowedOrigin := range h.config.CORSAllowedOrigins {
			if allowedOrigin == r.Header["Origin"][0] {
				return true
			}
		}
		return false
	}
}
```

**Review by nashqueue — DISMISSED** [review]
Thank you for the fast fix ! Approved after comments have been addressed.

File: rpc/json/ws.go
Comment (tzdybal) [inline rpc/json/ws.go:36]: Couldn't we reuse some existing, well tested function from some CORS library?
This is security-related function (prevents XSS) - I think we should reuse something.
Also, I'm not 100% sure about expected semantics of this function (see comments below).

Absolute minimum - cover this with tests.

File: rpc/json/ws.go
Comment (tzdybal) [inline rpc/json/ws.go:42]: What if `"*"` is not the first element of `allowedOrigings`?

File: rpc/json/ws.go
Comment (tzdybal) [inline rpc/json/ws.go:39]: Are you sure? Why not `false`? How can we assure that this empty value is actually allowed/correct/secure? (see: https://stackoverflow.com/questions/42239643/when-do-browsers-send-the-origin-header-when-do-browsers-set-the-origin-to-null/42242802)

File: rpc/json/ws.go
Comment (tzdybal) [reply rpc/json/ws.go:68]: I'm a big fan of inlining here, instead of creating a factory method to create a function.

File: rpc/json/handler.go
Comment (tzdybal) [inline rpc/json/handler.go:27]: Actually, we already have [`RPCConfig` type](https://github.com/rollkit/rollkit/blob/ea0dec6801c84db7333b24f9d5c022ccc68557ba/config/rpc.go#L4) to avoid using tendermint types directly deep in rollkit.
There's also potential to extract a method from: https://github.com/rollkit/rollkit/blob/ea0dec6801c84db7333b24f9d5c022ccc68557ba/conv/config.go#L21-L29).

**Review by tzdybal — CHANGES_REQUESTED** [review]

File: rpc/json/ws.go
Comment (tuxcanfly) [reply rpc/json/ws.go]: I don't think there is any XSS vulnerability here since there is no requirement for a browser origin for the websocket rpc client. In fact, I think we can take the same approach as tendermint and accept _all_ origins, headers and methods without any security issue as such:

https://github.com/tendermint/tendermint/blob/35581cf54ec436b8c37fabb43fdaa3f48339a170/rpc/jsonrpc/server/ws_handler.go#L49

**Review by tuxcanfly — COMMENTED** [review]
Addressed review.

File: rpc/json/ws.go
Comment (tuxcanfly) [reply rpc/json/ws.go]: Yeah I thought about this but took a middle way between tendermint's _very_ permissive approach and a strict verification approach

https://github.com/tendermint/tendermint/blob/35581cf54ec436b8c37fabb43fdaa3f48339a170/rpc/jsonrpc/server/ws_handler.go#L49

I'm not sure there is a XSS vulnerability here as we are not expecting the RPC server to be restricted to a particular origin.

However the tendermint approach is misleading as it seems to accept the config variables only to never use them to restrict the origin.

Not sure what the right way to deal with this is, but happy to discuss ideas.

File: rpc/json/ws.go
Comment (gupadhyaya) [inline rpc/json/ws.go:53]: why do you need to check the individual values like `""` and `*`? can you not just do:
```
if h.config.IsCorsEnabled() {
	upgrader.CheckOrigin = func(r *http.Request) bool {
		for _, allowedOrigin := range h.config.CORSAllowedOrigins {
			if allowedOrigin == r.Header["Origin"][0] {
				return true
			}
		}
		return false
	}
}
```

Comment (tuxcanfly) [reply rpc/json/ws.go:53]: Yep, we can do this, but just need it to handle case where origin is not set, so that this doesn't throw index out of range.

Comment (tuxcanfly) [reply rpc/json/ws.go:53]: Commented above on @tzdybal review about how strict or permissive policy we should decide.

**Comment by tuxcanfly** [comment]
Suggestion, we should use tendermint's approach and use a permissive approach. So:

`CheckOrigin:     func(r *http.Request) bool { return true },`

Also we should remove the config options `CORSAllowedOrigins CORSAllowedMethods CORSAllowedHeaders` so that we don't have misleading options since we are _not_ using these options at all.

**Comment by gupadhyaya** [comment]
> CORSAllowedOrigins

agree. let's proceed with default pessimistic option like tendermint. not sure if we should remove the options thought. rollkit does not introduce new options, it is simply copying the CORs options from tendermint, which itself have not removed them. so let's keep the options as is.

**Comment by tuxcanfly** [comment]
Agreed, I realized these options are being shared with the RPC server, so agree on keeping them.

**Review by gupadhyaya — APPROVED** [review]

**Review by gupadhyaya — APPROVED** [review]

**Review by tzdybal — APPROVED** [review]

**Pull request merged**
